File: javascript/116.populating-next-right-pointers-in-each-node.js
/*
 * @lc app=leetcode id=116 lang=javascript
 *
 * [116] Populating Next Right Pointers in Each Node
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
  // BFS, keeping track of levels
  // O(n) time
  // O(n) space
  if (!root) return root
  let traversalOrder = []
  let queue = []
  queue.unshift([root, 0])
  while (queue.length > 0) {
    let vals = queue.shift()
    let currNode = vals[0]
    let level = vals[1]
    if (traversalOrder[level]) {
      traversalOrder[level].push(currNode)
    } else {
      traversalOrder[level] = [currNode]
    }
    if (currNode.left) {
      queue.unshift([currNode.left, level + 1])
    }
    if (currNode.right) {
      queue.unshift([currNode.right, level + 1])
    }
  }
  // Build next right pointers
  for (let i = 0; i < traversalOrder.length; i++) {
    for (let j = traversalOrder[i].length-1; j >= 1; j--) {
      traversalOrder[i][j].next = traversalOrder[i][j-1]
    }
  }
  return root
};
// @lc code=end

File: cpp/116.populating-next-right-pointers-in-each-node.cpp
/*
 * @lc app=leetcode id=116 lang=cpp
 *
 * [116] Populating Next Right Pointers in Each Node
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution
{
public:
    struct QueueVals
    {
        Node *root;
        int level;
    };
    Node *connect(Node *root)
    {
        // BFS traversal
        // keep track of all nodes at every level
        // iterate over nodes to create linked list
        // O(n) time O(n) space
        if (!root)
            return root;

        queue<QueueVals> myQueue;
        myQueue.push(QueueVals{root, 0});
        vector<vector<Node *>> levelNodes;
        while (myQueue.size() > 0)
        {
            QueueVals vals = myQueue.front();
            Node *currNode = vals.root;
            int currLevel = vals.level;
            myQueue.pop();
            if (currLevel == levelNodes.size())
            {
                // create new vector
                vector<Node *> newVect{currNode};
                levelNodes.push_back(newVect);
            }
            else
            {
                // push values to existing vector
                levelNodes[currLevel].push_back(currNode);
            }

            if (currNode->left)
            {
                myQueue.push(QueueVals{currNode->left, currLevel + 1});
            }
            if (currNode->right)
            {
                myQueue.push(QueueVals{currNode->right, currLevel + 1});
            }
        }

        // build linked lists at every level
        for (vector<Node *> nodes : levelNodes)
        {
            for (int i = 0; i < nodes.size() - 1; i++)
            {
                nodes[i]->next = nodes[i + 1];
            }
        }

        return root;
    }
};
// @lc code=end
